### PR TITLE
[MSPAINT] Establish Undo/Redo management

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Providing rich and consistent UI/UX to the users.
JIRA issue: [CORE-17969](https://jira.reactos.org/browse/CORE-17969)

## Proposed changes

- Painting the canvas is done by overlaying the multiple layers. Drawing each overlay is implemented as polymorphism of OOP.
- Refine the Undo/Redo mechanism.
- Some adjustments.

## TODO

- [x] Undo/Redo
- [x] VK_ESCAPE
- [x] Pen
- [x] Brush
- [x] Fill
- [x] Rect
- [x] RectSel
- [x] FreeSel
- [x] Eraser
- [x] Text
- [x] Polygon

### CODE REVIEW

File: base/applications/mspaint/canvas.h [hunk 1]
```diff
@@ -55,7 +56,8 @@ class CCanvasWindow : public CWindowImpl<CCanvasWindow>
     CANVAS_HITTEST m_hitSelection;
     CANVAS_HITTEST m_whereHit;
     POINT m_ptOrig; // The origin of drag start
-    HBITMAP m_hbmCached; // The cached buffer bitmap
+    HBITMAP m_hbmCached1; // The cached buffer bitmap #1
+    HBITMAP m_hbmCached2; // The cached buffer bitmap #2
```
Comment (HBelusca) [inline base/applications/mspaint/canvas.h:60]: hmmm... or array: `HBITMAP m_hbmCached[2];` ?

File: base/applications/mspaint/toolsmodel.cpp [hunk 2]
```diff
@@ -28,8 +28,8 @@ ToolsModel::ToolsModel()
 
 ToolsModel::~ToolsModel()
 {
-    for (size_t i = 0; i < TOOL_MAX + 1; ++i)
-        delete m_tools[i];
+    for (auto item : m_tools)
+        delete item;
```
Comment (HBelusca) [inline base/applications/mspaint/toolsmodel.cpp:32]:
```suggestion
    for (size_t i = 0; i < _countof(m_tools); ++i)
        delete m_tools[i];
```

File: base/applications/mspaint/winproc.cpp [hunk 3]
```diff
@@ -668,10 +673,19 @@ LRESULT CMainWindow::OnCommand(UINT nMsg, WPARAM wParam, LPARAM lParam, BOOL& bH
             imageModel.Redo();
             break;
         case IDM_EDITCOPY:
+            // FIXME: We should use CF_DIB in the future
```
Comment (HBelusca) [inline base/applications/mspaint/winproc.cpp:676]: Do you mean, CF_DIB *in addition* to CF_BITMAP? Or just CF_DIB? (some apps, maybe MS Paint too (but to be verified), save data in clipboard in more than 1 format)
Or, will copying in CF_DIB automatically perform conversion to CF_BITMAP and make that other format also available in clipboard?

File: base/applications/mspaint/mouse.cpp [hunk 4]
```diff
@@ -197,44 +240,89 @@ struct RectSelTool : ToolBase
...
+        m_bLeftButton = bLeftButton;
+        m_bDrawing = TRUE;
+        start.x = last.x = x;
+        start.y = last.y = y;
+        imageModel.NotifyImageChanged();
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y) override
+    {
+        last.x = x;
```
Comment (HBelusca) [inline base/applications/mspaint/mouse.cpp:295]: where is `last` declared? (and `start` in the function above) Is it a member of ToolBase? If so, why not using the usual `m_` prefix for these?

Comment (katahiromz) [reply base/applications/mspaint/mouse.cpp:295]: They are global variables. Shall I add `g_` prefix?

File: base/applications/mspaint/canvas.h (hunk 1 above)
Comment (katahiromz) [reply base/applications/mspaint/canvas.h]: Renamed and make them an array as `m_ahbmCached[2]`.

File: base/applications/mspaint/mouse.cpp (hunk 4 above)
Comment (katahiromz) [reply base/applications/mspaint/mouse.cpp]: Added. Renamed some global variables.

**Comment by katahiromz** [comment]
Ready.

**Pull request merged**